File: 3579.minimum-steps-to-convert-string-with-operations.java
#
# @lc app=leetcode id=3579 lang=java
#
# [3579] Minimum Steps to Convert String with Operations
#

# @lc code=start
class Solution {
    public int minOperations(String word1, String word2) {
        int n = word1.length();
        int INF = 201;
        int[] dp = new int[n + 1];
        for (int i = 0; i <= n; i++) {
            dp[i] = INF;
        }
        dp[0] = 0;
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < i; j++) {
                int c = getCost(word1, word2, j, i - 1);
                if (dp[j] < INF) {
                    dp[i] = Math.min(dp[i], dp[j] + c);
                }
            }
        }
        return dp[n];
    }

    private int getCost(String s, String t, int l, int r) {
        int m = r - l + 1;
        if (m == 0) return 0;
        // no reverse, swaps
        int[][] cnt = new int[26][26];
        int f = 0;
        for (int p = l; p <= r; p++) {
            int x = s.charAt(p) - 'a';
            int y = t.charAt(p) - 'a';
            cnt[x][y]++;
            if (x == y) f++;
        }
        int k = 0;
        for (int x = 0; x < 26; x++) {
            for (int y = x + 1; y < 26; y++) {
                k += Math.min(cnt[x][y], cnt[y][x]);
            }
        }
        int ops1 = m - f - k;
        // rev + swaps after (s_rev vs t)
        int[][] cnt_rev = new int[26][26];
        int f_rev = 0;
        for (int i = 0; i < m; i++) {
            int sp = l + (m - 1 - i);
            int tp = l + i;
            int x = s.charAt(sp) - 'a';
            int y = t.charAt(tp) - 'a';
            cnt_rev[x][y]++;
            if (x == y) f_rev++;
        }
        int k_rev = 0;
        for (int x = 0; x < 26; x++) {
            for (int y = x + 1; y < 26; y++) {
                k_rev += Math.min(cnt_rev[x][y], cnt_rev[y][x]);
            }
        }
        int ops3 = m - f_rev - k_rev + 1;
        // rev + swaps before (s vs t_rev)
        int[][] cnt_pre = new int[26][26];
        int f_pre = 0;
        for (int i = 0; i < m; i++) {
            int sp = l + i;
            int tp = l + (m - 1 - i);
            int x = s.charAt(sp) - 'a';
            int y = t.charAt(tp) - 'a';
            cnt_pre[x][y]++;
            if (x == y) f_pre++;
        }
        int k_pre = 0;
        for (int x = 0; x < 26; x++) {
            for (int y = x + 1; y < 26; y++) {
                k_pre += Math.min(cnt_pre[x][y], cnt_pre[y][x]);
            }
        }
        int ops4 = m - f_pre - k_pre + 1;
        int ans = ops1;
        ans = Math.min(ans, ops3);
        ans = Math.min(ans, ops4);
        return ans;
    }
}
# @lc code=end

File: 3579.minimum-steps-to-convert-string-with-operations.go
#
# @lc app=leetcode id=3579 lang=golang
#
# [3579] Minimum Steps to Convert String with Operations
#

# @lc code=start
func minOperations(word1 string, word2 string) int {
	n := len(word1)
	dp := make([]int, n+1)
	const INF = 1 << 30
	for i := range dp {
		dp[i] = INF
	}
	dp[0] = 0

	rev_string := func(s string) string {
		b := []byte(s)
		i, j := 0, len(b)-1
		for i < j {
			b[i], b[j] = b[j], b[i]
			i++
			j--
		}
		return string(b)
	}

	match_cost := func(a, b string) int {
		k := len(a)
		cnt := [26][26]int{}
		d := 0
		for p := 0; p < k; p++ {
			if a[p] != b[p] {
				x := int(a[p] - 'a')
				y := int(b[p] - 'a')
				cnt[x][y]++
				d++
			}
		}
		m := 0
		for x := 0; x < 26; x++ {
			for y := x+1; y < 26; y++ {
				times := cnt[x][y]
				if times > cnt[y][x] {
					times = cnt[y][x]
				}
				m += times
			}
		}
		return d - m
	}

	calc_cost := func(l, r int) int {
		k := r - l + 1
		s1 := word1[l : l+k]
		s2 := word2[l : l+k]
		c1 := match_cost(s1, s2)
		r1 := rev_string(s1)
		c2 := match_cost(r1, s2) + 1
		r2 := rev_string(s2)
		c3 := match_cost(s1, r2) + 1
		minC := c1
		if c2 < minC {
			minC = c2
		}
		if c3 < minC {
			minC = c3
		}
		return minC
	}

	for i := 1; i <= n; i++ {
		for j := 0; j < i; j++ {
			if dp[j] != INF {
				c := calc_cost(j, i-1)
				if dp[j] + c < dp[i] {
					dp[i] = dp[j] + c
				}
			}
		}
	}

	return dp[n]
}
# @lc code=end